horovod/ray/runner.py

```diff
@@ -0,0 +1,445 @@
+        return func(self.executable)
+
+
+@ray.remote
+class NodeColocator:
+    """Responsible for colocation of child actors.
+
+    These actors are given resources equal to the sum of resources
+    to be effectively allocated to their children. The child
+    workers are currently allocated 0 resources in this implementation.
+
+    This is a mechanism for gang-scheduling and could be replaced
+    later on with placement groups. Gang-scheduling must occur because
+    otherwise another concurrent group could be placed on this node.
+
+    Right now, the only resources that are explicitly propogated to
+    underlying colocated workers are cuda visible devices.
+
+    Args:
+        node_rank (int): Rank of the node that this colocator is placed on.
+        num_slots (int): Total number of slots on this machine.
+        world_size (int): Total number of workers (slots) participating
+            in the job across all nodes.
+        use_gpu (bool): Whether to utilize the GPUs on the node.
+    """
+
+    def __init__(self, *, node_rank: int, num_slots: int, world_size: int,
+                 use_gpu: bool):
+        self.node_rank = node_rank
+        self.num_slots = num_slots
+        self.world_size = world_size
+        if use_gpu:
+            gpu_ids = ray.get_gpu_ids()
+            assert len(gpu_ids) == num_slots, gpu_ids
+        self.workers = []
+
+    def create_workers(self,
+                       executable_cls: type = None,
+                       executable_args: list = None,
+                       executable_kwargs: dict = None):
+        """Colocates a number of workers.
+
+        Also passes on the CUDA_VISIBLE_DEVICES to each worker.
+
+        Args:
+            executable_cls (type): Class of object to be created on all
+                workers.
+            executable_args (list): Initialization arguments for the
+                executable_cls.
+            executable_kwargs (dict): Initialization arguments for the
+                executable_cls.
+        """
+        # Create a node ip resource label so that we can pin
+        # all of the child actors to the same node. This ensures
+        # colocation and balanced training.
+        node_id = f"node:{ray.services.get_node_ip_address()}"
+        remote_cls = ray.remote(BaseHorovodWorker)
+        remote_cls = remote_cls.options(
+            num_cpus=0, num_gpus=0, resources={node_id: 0.01})
+
+        world_rank_start = self.num_slots * self.node_rank
+
+        self.workers = [
+            remote_cls.remote(world_rank=rank, world_size=self.world_size)
+            for rank in range(world_rank_start, world_rank_start +
+                              self.num_slots)
+        ]
+
+        # Propogate cuda visible devices to the underlying
+        # colocated workers.
+        gpu_ids = ray.get_gpu_ids()
+        for worker, gpu_id in zip(self.workers, gpu_ids):
+            worker.update_env_vars.remote({"CUDA_VISIBLE_DEVICES": gpu_id})
```

Comment:
`gpu_id` is only one GPU, is that correct?

For NCCL, we need to avoid setting `CUDA_VISIBLE_DEVICES`.  NCCL needs to have visibility of adjacent devices in order to use CUDA IPC.  Typically, we handle GPU isolation at the framework level, for example:

```
torch.cuda.set_device(hvd.local_rank())
```

If you need to set `CUDA_VISIBLE_DEVICES` because multiple Ray applications may be using the same nodes, then we can set `"CUDA_VISIBLE_DEVICES": gpu_ids` so every worker can see the devices in its local communicator.